### PR TITLE
Bug fix for VIF loss and measure

### LINKED ISSUE #89: VIF returns nan
Similar to #86, if I put all zeros to VIF, it returns nan

```
In [1]: from photosynthesis_metrics import vif_p

In [2]: import torch

In [3]: x = torch.zeros(1, 3, 256, 256)

In [4]: vif_p(x, x)
/home/ostyakov/.local/lib/python3.6/site-packages/torch/tensor.py:330: UserWarning: non-inplace resize is deprecated
  warnings.warn("non-inplace resize is deprecated")
Out[4]: tensor([nan])
```

### PR DESCRIPTION
**Pull request by zakajd**
Closes #89 

## Proposed Changes
  - Test to cover bug behaviour
  - Fix to add EPS constant before division to avoid NaN
  - Some checks from [original MATLAB implementation](https://live.ece.utexas.edu/research/Quality/VIF.htm) that were missing
- `Reduction` parameter, to fix another small bug I found

### CODE REVIEW

File: photosynthesis_metrics/vif.py [hunk 1]
```diff
@@ -30,17 +30,18 @@ def _gaussian_kernel2d(kernel_size: int = 5, sigma: float = 2.0) -> torch.Tensor
     return kernel
 
 
-def vif_p(prediction: torch.Tensor, target: torch.Tensor,
-          sigma_n_sq: float = 2.0, data_range: Union[int, float] = 1.0) -> torch.Tensor:
+def vif_p(prediction: torch.Tensor, target: torch.Tensor, sigma_n_sq: float = 2.0,
```
Comment (denproc) [inline photosynthesis_metrics/vif.py:33]: Please, add input validation and adjustments to `vif_p`. As a result, there will be no need in validation check in `VIFLoss().forward()`, while `VIFLoss().compute_metric()`, which consists of one string, can be moved to `forward()`.

**Comment by zakajd** [comment]
@denproc  ready for review.
Thanks for idea to check loss computation. Found and fixes a bug with few in-place operations which prevented correct back propagation

File: photosynthesis_metrics/vif.py [hunk 2]
```diff
@@ -89,17 +92,38 @@ def vif_p(prediction: torch.Tensor, target: torch.Tensor,
...
 
-    return prediction_vif / target_vif
+    # Reduce if needed
+    if reduction == 'mean':
+        return vif.mean(dim=0)
+    elif reduction == 'sum':
+        return vif.sum(dim=0)
+    elif reduction != 'none':
+        raise ValueError(f'Expected reduction modes are "mean"|"sum"|"none", got {reduction}')
+    return vif
```
Comment (snk4tr) [inline photosynthesis_metrics/vif.py:126]: Consider reimplmeneting it [the way it is done in BRISQUE](https://github.com/photosynthesis-team/photosynthesis.metrics/blob/83e65bb48e0e61b6b89ae75aaa3fcbb63b74dac5/photosynthesis_metrics/brisque.py#L178)

**Comment by zakajd** [comment]
I also forgot to add check for correct input size. Because of progressive downsampling VIF can fail for small images.

File: tests/test_vif.py [hunk 3]
```diff
@@ -26,38 +26,99 @@ def target_1d() -> torch.Tensor:
...
-        vif_p(prediction_1d, target_1d, data_range=1.)
-    except Exception as e:
-        pytest.fail(f"Unexpected error occurred: {e}")
+    vif_p(prediction_1d, target_1d, data_range=1.)
+
+
+def test_vif_p_one_for_equal_tensors(prediction: torch.Tensor) -> None:
+    target = prediction.clone()
+    measure = vif_p(prediction, target)
+    print(measure)
```
Comment (denproc) [inline tests/test_vif.py:39]: Remove `print`

File: photosynthesis_metrics/vif.py [hunk 4]
```diff
@@ -51,7 +52,13 @@ def vif_p(prediction: torch.Tensor, target: torch.Tensor,
         See https://live.ece.utexas.edu/research/Quality/VIF.htm for details.
         
     """
-    assert prediction.dim() == 4, f'Expected 4D tensor, got {prediction.size()}'
+    _validate_input((prediction, target), allow_5d=False)
+    prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+
+    min_size = 41
```
Comment (denproc) [inline photosynthesis_metrics/vif.py:58]: Please, add comment how to get this number.

Comment (zakajd) [reply photosynthesis_metrics/vif.py:58]: That's magic 🧙 
Due to progressive downsampling and different kernel sizes on different levels I didn't find formula to correctly estimate min size. I just tried a few different resolutions

**Comment by zakajd** [comment]
@denproc  please review and merge if everything OK
